Documentation/proposals/202411-zone-aware-sharding.md

```diff
@@ -0,0 +1,311 @@
+
+Given these values, a target is to be scraped by a given prometheus instance
+by using the following algorithm:
+
+```go
+assert(num_shards >= len(zones))     # Error: zone(s) will not get scraped
+assert(num_shards % len(zones) == 0) # Warning: multi-scraping of instances
+
+shards_per_zone := max(1, floor(num_shards / len(zones)))
+prom_zone_idx := shard_index % shards_per_zone
+    
+if zone_label == zones[prom_zone_idx] {
+    
+    assignment_idx := floor(shard_index / shards_per_zone)
+
+    if hash(address) % shards_per_zone == assignment_idx {
+        do_scrape()
+    }
+}
+```
+
+By using modulo to calculate the `prom_zone_idx`, instances will be distributed
+to zones in the sense of `A,B,C,A,B,C` and so on. This allows a modification of
+the `num_shards` value without redistribution of shards or data.
+This was preferred over allowing the number of `zones` to change, as this is
+less likely to happen.
+
+### Edge cases
+
+We have introduced asserts in the above section to warn about edge cases that
+might lead to duplicate data or data loss.
+
+By the above algorithm. prometheus instances will be distributed in an
+alternating fashion by using the already existing shard index.
+This leads to the following edge cases:
+
+When `num_shards` is 10 and `len(zones)` is 3 as in `[A..C]`.
+`shards_per_zone` is 3. This yields the following distribution
+
+```text
+0 1 2 | 3 4 5 | 6 7 8 | 9 | shard index
+A B C | A B C | A B C | A | zone
+0 0 0 | 1 1 1 | 2 2 2 | 0 | assignment index
+```
+
+In this case the 2nd assert will warn about double scraping of instances in
+zone A, as the same targets are being assigned to instance 0 and 9.
+
+When `num_shards` is 2 and `len(zones)` is 3 as in `[A..C]`.
+`shards_per_zone` is 1. This yields the following distribution
+
+```text
+0 1   | shard index
+A B C | zone
+0 0 0 | assignment index
+```
+
+In this case the 1st assert will warn about zone C not being scraped.
+
+This second case - a zone not being scraped - should lead to an error in the
+operator, causing the change to not be rolled out, as data would be lost.
+The first case - double scraping - should at minimum cause a warning.
+
+## Required configuration options
+
+Following the algorithm presented above, we suggest the following configuration
+options to be added to the [Prometheus custom resource definition](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#monitoring.coreos.com/v1.Prometheus).
+
+All values used in this snippet should also be the defaults for their
+corresponding keys.
+
+```yaml
+spec:
+  shardingStrategy:
+     # Select a sharding mode. Can be 'Classic' or 'Topology'
+    mode: 'Classic'    
+
+    # Optional: array of relabel configurations to insert before the sharding
+    # relabel configuration.
+    additionalRelabelConfig: [] 
```

Comment:
Scrape classes can also be leveraged to inject relabel configs. While it's more explicit to have a dedicated field here, it duplicates the same functionality. I'd rather avoid adding a new field for now and document the possibility to use scrape classes.

Also we could already provide more flexibility to users by letting them set the __tmp_hash label before the hashmod rule.E.g.  instead of having this snippet at the end of the generated relabel configs:

````yaml
relabel_configs:
  - ...
  - source_labels: [__address__]
    separator: ;
    modulus: 2
    target_label: __tmp_hash
    replacement: $1
    action: hashmod
  - source_labels: [__tmp_hash]
    separator: ;
    regex: "0"
    replacement: $1
    action: keep
````

It could be

```yaml
relabel_configs:
  - ...
  # If the user has set __tmp_hash before then use this value otherwise default to __address__.
  - source_labels: [__address__, __tmp_hash]
    target_label: __tmp_hash
    regex: "(.+);(.+)"
    replacement: $1
    action: replace
  - source_labels: [__tmp_hash]
    separator: ;
    modulus: 2
    target_label: __tmp_hash
    replacement: $1
    action: hashmod
  - source_labels: [__tmp_hash]
    separator: ;
    regex: "0"
    replacement: $1
    action: keep
````

Reply:
I removed the field and added a note on scrape classes in the "alternatives discussed" section.

Reply:
> If scrape classes are always injected before this block and essentially provide the same functionality as additionalRelabelConfig we can remove that field.

the API [documentation](https://prometheus-operator.dev/docs/api-reference/api/#monitoring.coreos.com/v1.ScrapeClass) says:

> The Operator automatically adds relabelings for a few standard Kubernetes fields like __meta_kubernetes_namespace and __meta_kubernetes_service_name. Then the Operator adds the scrape class relabelings defined here. Then the Operator adds the target-specific relabelings defined in the scrape object.

Reply:
As the field is already gone -> resolved

Reply:
https://github.com/prometheus-operator/prometheus-operator/pull/7126, let's see if it breaks the e2e tests (it shouldn't).